src/Symfony/Component/Config/Definition/Builder/TreeBuilder.php

```diff
@@ -16,13 +16,18 @@
 /**
  * This is the entry class for building a config tree.
  *
+ * @template T of 'variable'|'scalar'|'boolean'|'integer'|'float'|'enum'|'array'
+ *
  * @author Johannes M. Schmitt <schmittjoh@gmail.com>
  */
 class TreeBuilder implements NodeParentInterface
 {
     protected ?NodeInterface $tree = null;
     protected ?NodeDefinition $root = null;
 
+    /**
+     * @param T $type
+     */
     public function __construct(string $name, string $type = 'array', ?NodeBuilder $builder = null)
```

Comment:
Not so happy with this, because technically, it is incorrect, `new TreeBuilder('x')` would result in `TreeBuilder<'variable'|'scalar'|'boolean'|'integer'|'float'|'enum'|'array'>`, so `getRootNode` would return `VariableNodeDefinition|ScalarNodeDefinition|....`, but can't really fix it without a BC break.

However, those specifying the second argument will get better return type, everyone else will still get basically `NodeDefinition|ArrayNodeDefinition` which is same as before.